File: bindings/kotlin/src/main/kotlin/dev/tachyon_ipc/RpcBus.kt
package dev.tachyon_ipc

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.cancellable
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.yield

/**
 * Bidirectional RPC bus backed by two independent SPSC arenas.
 *
 * Use [listen] on the callee side and [connect] on the caller side.
 * Always call [close] when done, or wrap in a `use {}` block.
 *
 * Not thread-safe. Caller side ([call], [waitResponse]) and callee side
 * ([serve], [reply]) must each be accessed from a single coroutine.
 */
class RpcBus private constructor(private val inner: TachyonRpcBus) : AutoCloseable {

    companion object {
        /**
         * Creates two SHM arenas and blocks until a caller connects via [path].
         * Interrupted syscalls (EINTR) are retried transparently.
         *
         * @param path    UNIX socket path.
         * @param capFwd  Arena capacity for incoming requests. Must be a power of two.
         * @param capRev  Arena capacity for outgoing replies. Must be a power of two.
         */
        fun listen(path: String, capFwd: Long, capRev: Long): RpcBus =
            RpcBus(TachyonRpcBus.rpcListen(path, capFwd, capRev))

        /**
         * Attaches to existing SHM arenas via the UNIX socket at [path].
         *
         * @throws TachyonException with code 14 on ABI mismatch.
         */
        fun connect(path: String): RpcBus =
            RpcBus(TachyonRpcBus.rpcConnect(path))
    }

    /**
     * Configures the consumer polling back-off strategy for both arenas.
     *
     * [pureSpin] = 1 enables pure-spin mode; 0 restores hybrid futex-sleep mode.
     * Call before the first message.
     */
    fun setPollingMode(pureSpin: Int) {
        inner.setPollingMode(pureSpin)
    }

    /**
     * Copies [payload] into arena_fwd and suspends until the response matching
     * the assigned correlation ID arrives in arena_rev.
     *
     * Yields the coroutine if the TX ring buffer is full.
     * Returns the response as an [RpcMessage] with the correlation ID echoed.
     *
     * @param payload       Request bytes.
     * @param msgType       Application-level message type (uint16 range).
     * @param spinThreshold Spin iterations before futex sleep on the RX side.
     */
    suspend fun call(
        payload: ByteArray,
        msgType: Int,
        spinThreshold: Int = 10_000,
    ): RpcMessage {
        var cid: Long
        while (true) {
            try {
                cid = inner.call(payload, msgType)
                break
            } catch (e: BufferFullException) {
                yield()
            }
        }
        return waitResponse(cid, spinThreshold)
    }

    /**
     * Sends [payload] into arena_fwd without waiting for a response.
     * Returns the assigned correlation ID for use with [waitResponse].
     *
     * Yields the coroutine if the TX ring buffer is full.
     *
     * @param payload Request bytes.
     * @param msgType Application-level message type (uint16 range).
     */
    suspend fun send(payload: ByteArray, msgType: Int): Long {
        while (true) {
            try {
                return inner.call(payload, msgType)
            } catch (e: BufferFullException) {
                yield()
            }
        }
    }

    /**
     * Suspends until the response matching [correlationId] arrives in arena_rev.
     * Returns the response as an [RpcMessage].
     *
     * Yields the coroutine while spinning.
     *
     * @param correlationId The ID returned by [send].
     * @param spinThreshold Spin iterations before futex sleep.
     * @throws PeerDeadException on correlation mismatch or fatal arena error.
     */
    suspend fun waitResponse(correlationId: Long, spinThreshold: Int = 10_000): RpcMessage {
        while (true) {
            inner.waitResponse(correlationId, spinThreshold).use { rx ->
                if (rx != null) {
                    return RpcMessage(
                        correlationId = rx.correlationId,
                        msgType = rx.msgType,
                        data = rx.getData(),
                    )
                }
            }
            yield()
        }
    }

    /**
     * Returns a [Flow] of inbound [RpcMessage] values from arena_fwd.
     *
     * Each emission carries the [RpcMessage.correlationId] required by [reply].
     * The flow runs on [Dispatchers.IO], yields between spin iterations,
     * and is cancellable.
     *
     * @param spinThreshold Spin iterations before yielding on each empty poll.
     */
    fun serve(spinThreshold: Int = 10_000): Flow<RpcMessage> = flow {
        while (true) {
            inner.serve(spinThreshold).use { rx ->
                if (rx != null) {
                    emit(
                        RpcMessage(
                            correlationId = rx.correlationId,
                            msgType = rx.msgType,
                            data = rx.getData(),
                        )
                    )
                } else {
                    yield()
                }
            }
        }
    }.flowOn(Dispatchers.IO).cancellable()

    /**
     * Copies [payload] into arena_rev as a response to [correlationId].
     *
     * The [RpcRxGuard] from [serve] must be committed (via `use {}` exit)
     * before calling this method.
     * Yields the coroutine if the TX ring buffer is full.
     *
     * @param correlationId The ID from the [RpcMessage] emitted by [serve].
     *                      Must be greater than zero.
     * @param payload       Response bytes.
     * @param msgType       Application-level response type (uint16 range).
     * @throws TachyonException if [correlationId] is zero or commit fails.
     */
    suspend fun reply(correlationId: Long, payload: ByteArray, msgType: Int) {
        while (true) {
            try {
                inner.reply(correlationId, payload, msgType)
                return
            } catch (e: BufferFullException) {
                yield()
            }
        }
    }

    override fun close() {
        inner.close()
    }
}

File: bindings/csharp/src/TachyonIpc/RpcBus.cs
using System.Runtime.CompilerServices;
using TachyonIpc.Native;

namespace TachyonIpc;

/// <summary>
/// Managed wrapper around a native <c>tachyon_rpc_bus_t</c>.
/// Not thread-safe: caller side (<see cref="Call"/>, <see cref="Wait"/>) and callee side
/// (<see cref="Serve"/>, <see cref="Reply"/>) must each be accessed from a single thread.
/// </summary>
public sealed unsafe class RpcBus : IDisposable
{
    private nint _rpc;

    static RpcBus() => NativeLoader.Register();

    private RpcBus(nint rpc) => _rpc = rpc;

    /// <summary>
    /// Creates two SHM arenas and blocks until a caller connects via <paramref name="socketPath"/>.
    /// EINTR is retried transparently.
    /// </summary>
    /// <param name="socketPath">UDS socket path. Unlinked automatically after handshake.</param>
    /// <param name="capFwd">arena_fwd capacity in bytes (caller to callee). Must be a power of two.</param>
    /// <param name="capRev">arena_rev capacity in bytes (callee to caller). Must be a power of two.</param>
    /// <exception cref="TachyonException">SHM creation or socket binding failure.</exception>
    public static RpcBus Listen(string socketPath, nuint capFwd, nuint capRev)
    {
        nint rpc;
        TachyonError err;
        do
        {
            err = TachyonNative.tachyon_rpc_listen(socketPath, capFwd, capRev, &rpc);
        } while (err == TachyonError.Interrupted);

        TachyonException.ThrowIfError(err, nameof(TachyonNative.tachyon_rpc_listen));
        return new RpcBus(rpc);
    }

    /// <summary>
    /// Attaches to existing SHM arenas via the UDS socket at <paramref name="socketPath"/>.
    /// </summary>
    /// <param name="socketPath">UDS socket path of the listener.</param>
    /// <exception cref="TachyonException">
    /// <see cref="TachyonError.AbiMismatch"/> if the callee was compiled with a different
    /// Tachyon version or <c>TACHYON_MSG_ALIGNMENT</c>.
    /// </exception>
    public static RpcBus Connect(string socketPath)
    {
        nint rpc;
        TachyonException.ThrowIfError(
            TachyonNative.tachyon_rpc_connect(socketPath, &rpc),
            nameof(TachyonNative.tachyon_rpc_connect));
        return new RpcBus(rpc);
    }

    /// <summary>
    /// Controls the polling back-off strategy for both arenas.
    /// Call before the first message.
    /// When <paramref name="pureSpin"/> is <c>true</c>, the producer skips the futex wake check
    /// on every flush. Only use this when both consumer threads are dedicated and never park.
    /// </summary>
    /// <param name="pureSpin"><c>true</c> for busy-spin, <c>false</c> for adaptive futex mode.</param>
    public void SetPollingMode(bool pureSpin)
    {
        ThrowIfDisposed();
        TachyonNative.tachyon_rpc_set_polling_mode(_rpc, pureSpin ? 1 : 0);
    }

    /// <summary>Current composite bus state. Returns <see cref="TachyonState.FatalError"/> if
    /// either arena has entered a fatal error state.</summary>
    public TachyonState State
    {
        get
        {
            ThrowIfDisposed();
            return TachyonNative.tachyon_rpc_get_state(_rpc);
        }
    }

    /// <summary>
    /// Copies <paramref name="payload"/> into arena_fwd and returns the assigned correlation ID.
    /// Blocks if the ring buffer is full.
    /// </summary>
    /// <param name="payload">Request payload.</param>
    /// <param name="msgType">Application-level message type (uint16 range).</param>
    /// <returns>Assigned correlation ID, always greater than zero.</returns>
    /// <exception cref="TachyonException">Commit failure or internal error.</exception>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public ulong Call(ReadOnlySpan<byte> payload, uint msgType)
    {
        ThrowIfDisposed();
        ulong cid;
        fixed (byte* p = payload)
        {
            TachyonException.ThrowIfError(
                TachyonNative.tachyon_rpc_call(_rpc, p, (nuint)payload.Length, msgType, &cid),
                nameof(TachyonNative.tachyon_rpc_call));
        }

        return cid;
    }

    /// <summary>
    /// Blocks until the response matching <paramref name="correlationId"/> arrives in arena_rev,
    /// spinning then falling back to futex sleep.
    /// A correlation ID mismatch triggers a fatal error on arena_rev.
    /// </summary>
    /// <param name="correlationId">The ID returned by <see cref="Call"/>.</param>
    /// <param name="spinThreshold">Spin iterations before futex sleep.</param>
    /// <returns>An <see cref="RpcRxGuard"/> holding the response payload.</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public RpcRxGuard Wait(ulong correlationId, uint spinThreshold = 10_000)
    {
        ThrowIfDisposed();
        nuint actualSize;
        uint typeId;
        var ptr = TachyonNative.tachyon_rpc_wait(_rpc, correlationId, &actualSize, &typeId, spinThreshold);
        if (ptr == null)
            throw new TachyonException(TachyonNative.tachyon_rpc_get_state(_rpc) == TachyonState.FatalError
                    ? TachyonError.System
                    : TachyonError.Interrupted,
                nameof(TachyonNative.tachyon_rpc_wait));
        return new RpcRxGuard(_rpc, ptr, actualSize, typeId, correlationId, isServe: false);
    }

    /// <summary>
    /// Blocks until a request arrives in arena_fwd, spinning then falling back to futex sleep.
    /// Commit the returned guard before calling <see cref="Reply"/> to avoid holding both
    /// arena slots simultaneously.
    /// </summary>
    /// <param name="spinThreshold">Spin iterations before futex sleep.</param>
    /// <returns>An <see cref="RpcRxGuard"/> holding the request payload and correlation ID.</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public RpcRxGuard Serve(uint spinThreshold = 10_000)
    {
        ThrowIfDisposed();
        ulong correlationId;
        uint typeId;
        nuint actualSize;
        var ptr = TachyonNative.tachyon_rpc_serve(_rpc, &correlationId, &typeId, &actualSize, spinThreshold);
        if (ptr == null)
            throw new TachyonException(TachyonNative.tachyon_rpc_get_state(_rpc) == TachyonState.FatalError
                    ? TachyonError.System
                    : TachyonError.Interrupted,
                nameof(TachyonNative.tachyon_rpc_serve));
        return new RpcRxGuard(_rpc, ptr, actualSize, typeId, correlationId, isServe: true);
    }

    /// <summary>
    /// Copies <paramref name="payload"/> into arena_rev as a response to <paramref name="correlationId"/>.
    /// The <see cref="RpcRxGuard"/> from <see cref="Serve"/> must be committed before calling this method.
    /// </summary>
    /// <param name="correlationId">Must match the value from the served <see cref="RpcRxGuard"/>.</param>
    /// <param name="payload">Response payload.</param>
    /// <param name="msgType">Application-level response type (uint16 range).</param>
    /// <exception cref="TachyonException">
    /// <see cref="TachyonError.InvalidSize"/> if <paramref name="correlationId"/> is zero.
    /// </exception>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Reply(ulong correlationId, ReadOnlySpan<byte> payload, uint msgType)
    {
        ThrowIfDisposed();
        fixed (byte* p = payload)
        {
            TachyonException.ThrowIfError(
                TachyonNative.tachyon_rpc_reply(_rpc, correlationId, p, (nuint)payload.Length, msgType),
                nameof(TachyonNative.tachyon_rpc_reply));
        }
    }

    /// <summary>Unmaps both SHM arenas and releases all resources. Idempotent.</summary>
    public void Dispose()
    {
        var rpc = Interlocked.Exchange(ref _rpc, nint.Zero);
        if (rpc != nint.Zero)
            TachyonNative.tachyon_rpc_destroy(rpc);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void ThrowIfDisposed()
    {
        if (_rpc == nint.Zero)
            throw new ObjectDisposedException(nameof(RpcBus));
    }
}